mlflow/pyfunc/__init__.py

```diff
@@ -380,52 +410,95 @@ def _enforce_schema(pdf: PyFuncInput, input_schema: Schema):
-        expected_names = set(col_names)
-        actual_names = set(pdf.columns)
-        missing_cols = expected_names - actual_names
-        extra_cols = actual_names - expected_names
+        expected_cols = set(col_names)
+        actual_cols = set()
+        if len(expected_cols) == 1 and isinstance(pfInput, np.ndarray):
+            # for schemas with a single column, match input with column
+            pfInput = {col_names[0]: pfInput}
+            actual_cols = expected_cols
+        elif isinstance(pfInput, pandas.DataFrame):
+            actual_cols = set(pfInput.columns)
+        elif isinstance(pfInput, dict):
+            actual_cols = set(pfInput.keys())
+        missing_cols = expected_cols - actual_cols
+        extra_cols = actual_cols - expected_cols
         # Preserve order from the original columns, since missing/extra columns are likely to
         # be in same order.
         missing_cols = [c for c in col_names if c in missing_cols]
-        extra_cols = [c for c in pdf.columns if c in extra_cols]
+        extra_cols = [c for c in actual_cols if c in extra_cols]
         if missing_cols:
             message = (
-                "Model input is missing columns {0}."
-                " Note that there were extra columns: {1}".format(missing_cols, extra_cols)
+                "Model is missing inputs {0}."
+                " Note that there were extra inputs: {1}".format(missing_cols, extra_cols)
             )
             raise MlflowException(message)
-    else:
+    elif not input_schema.is_tensor_spec():
         # The model signature does not specify column names => we can only verify column count.
-        if len(pdf.columns) < len(input_schema.columns):
+        num_actual_columns = len(pfInput.columns)
+        if num_actual_columns < len(input_schema.data_rep):
             message = (
-                "Model input is missing input columns. The model signature declares "
-                "{0} input columns but the provided input only has "
-                "{1} columns. Note: the columns were not named in the signature so we can "
+                "Model inference is missing inputs. The model signature declares "
+                "{0} inputs  but the provided value only has "
+                "{1} inputs. Note: the inputs were not named in the signature so we can "
                 "only verify their count."
-            ).format(len(input_schema.columns), len(pdf.columns))
+            ).format(len(input_schema.data_rep), num_actual_columns)
             raise MlflowException(message)
-        col_names = pdf.columns[: len(input_schema.columns)]
-    col_types = input_schema.column_types()
-    new_pdf = pandas.DataFrame()
-    for i, x in enumerate(col_names):
-        new_pdf[x] = _enforce_type(x, pdf[x], col_types[i])
-    return new_pdf
+
+    if input_schema.is_tensor_spec():
+        if input_schema.has_column_names():
+            if isinstance(pfInput, dict):
```

Comment:
1) if there is only a single input, we also want to support a numpy array and wrap it in a dictionary with the expected input name. 
2) pandas.DataFrame should be allowed: 
   if there are multiple inputs in the schema, we should treat it as a dict[column -> column.values].